Avoid using the interpreter from TDataMember if possible.

Teemperor opened the pull request:
We can't use the interpreter when generating a PCM as this would
generate AST nodes which then would end up in the module, which
is causing a long chain of modules (such as redefinitons as we
suddenly have the same cling warpper function multiple times).

In this code path we seem to always have a number that we want
to convert to a string. So let's just use atol instead here if
the argument is just a number, which should avoid the issue with
the generated code.

As we also now check if the input is a number, I added an assert
that verifies we only call atol when the string is actually a
number.

vgvassilev commented:
@pcanal ping.

Teemperor commented:
I have to update the PR, so undo ping at Philippe :)

vgvassilev commented:
@Teemperor, did you mean updating it by reducing the scope of the fix as we discussed before? I still would like @pcanal to comment on the general direction, esp that we know we have a single use case of this code in Eve, which is the constant `9` ;)

Axel-Naumann commented on core/meta/src/TDataMember.cxx:
```diff
@@ -402,17 +404,24 @@ void TDataMember::Init(bool afterReading)
             Long_t l=0;
             Int_t  *value;
             TGlobal *enumval = gROOT->GetGlobal(ptr1,kTRUE);
```
Why do you call `TROOT::GetGlobal()` if the string is just a number?

pcanal replied:
i.e. Should the strtol be even before this call?

Axel-Naumann commented on core/meta/src/TDataMember.cxx:
```diff
@@ -402,17 +404,24 @@ void TDataMember::Init(bool afterReading)
-            if (enumval){
+
+            char *strtolResult;
+            l = std::strtol(ptr1, &strtolResult, 10);
```
Can you move the decl of `l` here?

Axel-Naumann commented on core/meta/src/TDataMember.cxx:
```diff
@@ -402,17 +404,24 @@ void TDataMember::Init(bool afterReading)
                if (obj)
                   l = ((TEnumConstant*)obj)->GetValue();
-               else
+               else if (!isnumber)
```
`isnumber` must be false here, right?

pcanal commented:
> So let's just use atol instead here if
the argument is just a number, which should avoid the issue with
the generated code.

Yes, checking 'early' that the string is just a number is a good idea.

Teemperor commented:
@Axel-Naumann the assert failure was for a previous version of this PR, the one you reviewed already had the fix :)

I think I addressed all comments in the newest revision, let's see if we get a green Jenkins with the current state of the master.

vgvassilev commented:
@Teemperor I hate to say this but do you feel like adding a gtest ;) ?

Axel-Naumann commented:
@pcanal can the be merged?

vgvassilev commented:
@Teemperor could you fix the coding violations?

pcanal commented:
@Axel-Naumann Yes (pending clang-formating).

Teemperor commented:
Back from vacation, formatted the code.
